### PR TITLE
GEP-2895: Query Parameter Filter

### PR DESCRIPTION
**Pull request by lianglli**
**What type of PR is this?**
/kind gep

**What this PR does / why we need it**:
The `HTTPRouteFilter` API now supports filters `RequestHeaderModifier` and `ResponseHeaderModifier`. 
This GEP proposes adding support for modifying query parameters in a `HTTPRoute`.

**Which issue(s) this PR fixes**:
Fixes #[2895](https://github.com/kubernetes-sigs/gateway-api/issues/2895)

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

**Comment by robscott** [comment]
Hey @lianglli, we're still very focused on getting v1.1 out the door so won't have time for a thorough review until after that's released. With that said, I'd recommend adding a section that shows which underlying implementations would support this. As far as I can tell, NGINX would and Envoy would not. It would be worth confirming that + looking at other common dataplanes like HAProxy to understand how widely implementable an API like this would be.

**Comment by costinm** [comment]
I was wondering: do we have WG participants who work on or have expertise in Nginx, HAProxy, Apache, caddy, traefik and the other data plane implementations ? I know Envoy is over represented, so not a problem - but it would be worth having some wiki page with various data planes - and few names that can provide review/feedback. 

And maybe a common checklist for each proposal - to formalize a bit more the process of making sure the features and 
APIs we add are implementable.

**Comment by spacewander** [comment]
> I was wondering: do we have WG participants who work on or have expertise in Nginx, HAProxy, Apache, caddy, traefik and the other data plane implementations ? I know Envoy is over represented, so not a problem - but it would be worth having some wiki page with various data planes - and few names that can provide review/feedback.
> 
> And maybe a common checklist for each proposal - to formalize a bit more the process of making sure the features and APIs we add are implementable.

I would like to help in this WG. Both Envoy and Nginx are my specialties.

I was the core developer of Apache APISIX, which is a popular gateway built upon Nginx: 
https://github.com/apache/apisix
https://github.com/apache/apisix/graphs/contributors.

Nowadays, I am working on Envoy-based gateway. I have contributed some code to Envoy: https://github.com/envoyproxy/envoy/commits?author=spacewander, especially on the Golang filter: https://github.com/envoyproxy/envoy/blob/d1e95536aec14330a4235ad888246868d686cfac/CODEOWNERS#L383

**Review by LiorLieberman — COMMENTED** [review]
> Hey @lianglli, we're still very focused on getting v1.1 out the door so won't have time for a thorough review until after that's released. With that said, I'd recommend adding a section that shows which underlying implementations would support this. As far as I can tell, NGINX would and Envoy would not. It would be worth confirming that + looking at other common dataplanes like HAProxy to understand how widely implementable an API like this would be.

AFAICT it is possible with lua filter in envoy (https://github.com/envoyproxy/envoy/issues/2098) and there is a [PR](https://github.com/envoyproxy/envoy/pull/33977) from 8 hours ago to support this on the Route level.

+1 to adding a section as Rob proposed though.

**Comment by lianglli** [comment]
> Hey @lianglli, we're still very focused on getting v1.1 out the door so won't have time for a thorough review until after that's released. With that said, I'd recommend adding a section that shows which underlying implementations would support this. As far as I can tell, NGINX would and Envoy would not. It would be worth confirming that + looking at other common dataplanes like HAProxy to understand how widely implementable an API like this would be.

Currently, both the KONG and Traefik support Query Parameter Modification based on its plugin.
And, it is natively supported by Tengine-Ingress.

File: geps/gep-2895/index.md [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+
+## TLDR
+
+Just like modify header is useful, the same goes for query parameters. 
+It's helpful to have a `HTTPQueryParamFilter` field in `HTTPRouteFilter` to set, 
+add and remove a query parameter of the HTTP request before it is sent to the upstream target.
+
+## Goals
+
+* Provide a way to modify query parameters in a `HTTPRoute`.
```
Comment (mlavacca) [inline geps/gep-2895/index.md:14]: I think it is worth mentioning here we are talking about transforming the request

File: geps/gep-2895/index.md [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+      type: QueryParamModifier
+    matches:
+    - queryParams:
+      - name: gray
+        type: Exact
+        value: 3
+
+```
+
+## Prior Art
```
Comment (mlavacca) [inline geps/gep-2895/index.md:160]:
```suggestion
## Implementation-specific solutions
```
I wouldn't use the "Prior art" term, as it is not a community-based prior art, but instead an implementation-specific effort.

File: geps/gep-2895/index.md [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+plugins:
+- name: request-transformer
+  route: my-route
+  config:
+    add:
+      querystring:
+      - new-param=some-value
+```
+
+### Traefik supports this with a Query Paramter Modification plugin
```
Comment (mlavacca) [inline geps/gep-2895/index.md:190]:
```suggestion
### Traefik supports this with a Query Parameter Modification plugin
```

File: geps/gep-2895/index.md [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+
+* This Traefik plugin allows user to modify the query parameters of an incoming request, by either adding new, deleting or modifying existing query parameters.
```
Comment (mlavacca) [inline geps/gep-2895/index.md:192]:
```suggestion
* This Traefik plugin allows users to modify the query parameters of an incoming request, by either adding new, deleting or modifying existing query parameters.
```

**Review by mlavacca — COMMENTED** [review]
Thanks for this PR, @lianglli!

**Comment by costinm** [comment]
> > I was wondering: do we have WG participants who work on or have expertise in Nginx, HAProxy, Apache, caddy, traefik and the other data plane implementations ? I know Envoy is over represented, so not a problem - but it would be worth having some wiki page with various data planes - and few names that can provide review/feedback.
> > And maybe a common checklist for each proposal - to formalize a bit more the process of making sure the features and APIs we add are implementable.
> 
> Pls. check the "## Prior Art" and "## References" of this PR specifically. Moreover, the `HTTPQueryParamFilter` is considered an extended feature.
> 
> BTW, I'm the core developer of Tengine and Tengine-Ingress.

I'm not familiar with Tengine - is it based on Nginx ? 

My question is more about having all "upstream" proxies that are used by different implementations ( nginx, envoy, all 'native' ones in rust/go/etc) - and info if it supports or not that feature. 

Adding optional features that are only supported by a few implementations is possible (I personally don't think it's right, but it's what this WG has decided), but I think there is a significant cost on the users and portability of the configs, so at least we should have the info.

File: geps/gep-2895/index.md [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+    // This must be an exact string match of query param name.
+    // (See https://www.rfc-editor.org/rfc/rfc7230#section-2.7.3).
+    //
+    // Input:
+    //   GET /foo?my-parameter=foo HTTP/1.1
+    //
+    // Config:
+    //   set:
+    //   - name: "my-parameter"
+    //     value: "bar"
```
[the remark this answers is not in the document]
Comment (lianglli) [reply geps/gep-2895/index.md:60]: This GEP proposes to add a new field `HTTPQueryParamFilter` to `HTTPRouteFilter`.

For the header is not set in the request, the `Set` operation of `HTTPHeaderFilter` does NOT describe it specifically.
However, the `Set` is same as the modify. And, the `Add` is same as the add and append. Lastly, `Remove` is same as the delete.

Hence, if the query parameter of `HTTPQueryParamFilter` is not set in the request, gateway will NOT do anything.

File: geps/gep-2895/index.md [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+    //
+    //
+    // +optional
+    // +listType=map
+    // +listMapKey=name
+    // +kubebuilder:validation:MaxItems=16
+    Set []HTTPHeader `json:"set,omitempty"`
+
+    // Add adds the given query param(s) (name, value) to the HTTP request
+    // before the action.
```
[the remark this answers is not in the document]
Comment (lianglli) [reply geps/gep-2895/index.md:73]: The `Set` operation of `HTTPHeaderFilter` is used to modify existing query params with the same name of an HTTP request. 

    // Input:
    //   GET /foo?my-parameter=foo HTTP/1.1
    //
    // Config:
    //   set:
    //   - name: "my-parameter"
    //     value: "bar"
    //
    // Output:
    //   GET /foo?my-parameter=bar HTTP/1.1

The PR has the above comments specifically.

File: geps/gep-2895/index.md (hunk 1 above)
Comment (lianglli) [reply geps/gep-2895/index.md:14]: Yes, it is.

File: geps/gep-2895/index.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (lianglli) [reply geps/gep-2895/index.md:73]: It is same as the `Set` action of `HTTPHeaderFilter`.

```yaml
type HTTPHeaderFilter struct {
    // Set overwrites the request with the given header (name, value)
    // before the action.
    //
    // Input:
    //   GET /foo HTTP/1.1
    //   my-header: foo
    //
    // Config:
    //   set:
    //   - name: "my-header"
    //     value: "bar"
    //
    // Output:
    //   GET /foo HTTP/1.1
    //   my-header: bar
    //
    // +optional
    // +listType=map
    // +listMapKey=name
    // +kubebuilder:validation:MaxItems=16
    Set []HTTPHeader `json:"set,omitempty"`
```

File: geps/gep-2895/index.md [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+    //     value: "bar"
+    //
+    // Output:
+    //   GET /foo HTTP/1.1
+    //
+    // +optional
+    // +listType=map
+    // +listMapKey=name
+    // +kubebuilder:validation:MaxItems=16
+    Set []HTTPHeader `json:"set,omitempty"`
```
Comment (spacewander) [inline geps/gep-2895/index.md:84]: There are two differences between Header and Query string:
1. Header is case insensitive while Query string is not.
2. The value of Header can not be empty but the one in Query string can be empty.

Would be better to use a new type.

Comment (lianglli) [reply geps/gep-2895/index.md:84]: OK. I'm checking it specifically.

File: geps/gep-2895/index.md [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+    //   GET /foo?my-parameter2=bar HTTP/1.1
+    //
+    // +optional
+    // +listType=set
+    // +kubebuilder:validation:MaxItems=16
+    Remove []string `json:"remove,omitempty"`
+}
+```
+
+## Examples
```
Comment (arkodg) [inline geps/gep-2895/index.md:128]: * suggest including all implementations outlined in https://gateway-api.sigs.k8s.io/implementations/#implementations . If they dont support it, NA or skipping it should be fine

File: geps/gep-2895/index.md (hunk 7 above)
Comment (lianglli) [reply geps/gep-2895/index.md:84]: @spacewander 

Based on the Gateway v1.0.0 and v1.1.0, the type `HeaderName` is the name of a header or query parameter.

So, it's better to commit a new PR about it.
However, this GEP-2895 will add specific comments about name and value of query parameter. 

```go
// HeaderName is the name of a header or query parameter.
//
// +kubebuilder:validation:MinLength=1
// +kubebuilder:validation:MaxLength=256
// +kubebuilder:validation:Pattern=`^[A-Za-z0-9!#$%&'*+\-.^_\x60|~]+$`
// +k8s:deepcopy-gen=false
type HeaderName string
```

Comment (spacewander) [reply geps/gep-2895/index.md:84]: I see

**Comment by robscott** [comment]
Hey @lianglli, we're working on scoping new features for v1.2, do you mind adding a comment to track support for this proposal in https://github.com/kubernetes-sigs/gateway-api/discussions/3103?

**Comment by lianglli** [comment]
> Hey @lianglli, we're working on scoping new features for v1.2, do you mind adding a comment to track support for this proposal in #3103?

Yes, I will add a comment about [GEP-2895: Query Parameter Filter](https://github.com/kubernetes-sigs/gateway-api/pull/2959#top) asap.

**Comment by shaneutt** [comment]
@lianglli checking in: where are we at with this one, what do you need in order to move forward?

**Comment by arkodg** [comment]
> @lianglli checking in: where are we at with this one, what do you need in order to move forward?

@shaneutt ptal at https://github.com/kubernetes-sigs/gateway-api/discussions/3103#discussioncomment-9616304

**Comment by shaneutt** [comment]
Awesome, thank you

**Pull request closed without merging**